### PR TITLE
Fixing some C++ conversion errors.

### PR DESCRIPTION
**Pull request by benvanik**
* Implicit conversion from int to bool.
* Implicit conversion from size_t to uint32_t.
* Implicit conversion from char* to uint8_t.

### CODE REVIEW

File: source/text_handler.cpp [hunk 1]
```diff
@@ -499,7 +499,7 @@ spv_result_t AssemblyContext::checkRangeAndIfHexThenSignExtend(
       // Hex values are a bit special. They decode as unsigned values, but
       // may represent a negative number.  In this case, the overflow bits
       // should be zero.
-      failed = (value & overflow_mask);
+      failed = (value & overflow_mask) ? true : false;
```
Comment (dmircevski) [inline source/text_handler.cpp]: This is pointless.  Does it solve a compilation problem?

File: source/validate_id.cpp [hunk 2]
```diff
@@ -417,7 +417,7 @@ bool idUsage::isValid<SpvOpTypeArray>(const spv_instruction_t* inst,
   } else if (5 == constInst->words.size()) {
     uint64_t value =
         constInst->words[3] | ((uint64_t)constInst->words[4]) << 32;
-    bool signedness = constResultType->second.inst->words[3];
+    bool signedness = constResultType->second.inst->words[3] ? true : false;
```
Comment (dmircevski) [inline source/validate_id.cpp]: Again, pointless.

Comment (benvanik) [inline source/validate_id.cpp]: I'd have not sent the pull if it were pointless :/

This occurs at warning level 3 in VC++, which is a common setting:
```
1>..\third_party\spirv-tools\source\validate_id.cpp(420): warning C4800: 'const unsigned int': forcing value to bool 'true' or 'false' (performance warning)
```
https://msdn.microsoft.com/en-us/library/b6801kcy.aspx

Irrespective of the reason the VC++ warning gives, implicitly casting a non-bool to a bool is not very readable. Were the value of the given word here (which is typed as uint32_t) be anything but 1 or 0 the behavior is hard to follow. Readability-wise, knowing the author intended to always treat any value not equal to zero as true is easier to parse from this (or related methods, such as !! commonly used in javascript or not not in python). Implicit casts from different width integer types is not great in modern C++ where bool has a very well-defined meaning and reliance on implicit or C-style casts is non-idiomatic.

Comment (dmircevski) [inline source/validate_id.cpp]: Oh, the implicit conversion still happens -- the `?` takes a `bool` for its first argument. :)

I find the readability argument less than convincing, given that this implicit conversion is a well-understood language feature dating back to the 80s.

I don't have a problem with accepting this particular change (especially since I agree with the other parts), but I would have a problem with a general rule requiring us to always write the silly `? true : false` everywhere just to suppress a single compiler's warning.

Comment (benvanik) [inline source/validate_id.cpp]: Yeah, I wish warnings cross-compiler were more consistent. Or there was a good way to get the superset of all warnings without playing whack-a-mole. Or that the C++ spec was well-defined enough such that all compilers would have the same spec-defined warnings (especially in cases like this, where you aren't even doing anything crazy!).

I think the readability argument is less about the conversion and more about when it happens.
```
if (words[3]) {  // clear to reader bool conversion is happening
bool foo = words[3] ? true : false;  // also clear bool conversion is happening
bool foo = words[3];  // not clear, requires understanding that words[3] is uint32_t but the semantics of the particular index 3 is boolean
```

When we were migrating to C++11 a coworker once came up with a pretty good rule of thumb: if switching your lvalue type to 'auto' requires an unfamiliar reader to leave the file to figure out what the type is your naming is unclear or you're missing a cast.

But, like anything readability-related: totally subjective :P

Comment (dmircevski) [inline source/validate_id.cpp]: How about `signedness = (words[3] != 0)`?

Comment (msc-) [inline source/validate_id.cpp]: If you wish to avoid implicit casts to bool, the correct fix is, e.g,
```
bool foo = words[3] > 0;
```
or, e.g,
```
failed = (value & overflow_mask) != 0;
```

Comment (benvanik) [inline source/validate_id.cpp]: Ooo I like the != 0 style - updated!

**Comment by dmircevski** [comment]
LGTM, merged.  Thanks, Ben!
